ygen/enumgen.go

```diff
@@ -587,49 +590,140 @@ func newEnumGenState() *enumGenState {
-				// different names.
+				return nil, fmt.Errorf("enumgen.go bug: enumKey %q has been given a second name during name generation, this should be detected and disallowed in the algorithm; one of the entries used for name generation: %+v", enumKey, nameClashSet[enumKey])
+			}
+		}
+
+		// If there is no clash, then we're done. This should be the vast majority of cases.
+		if len(nameClashSet) == 1 {
+			var enumKey string
+			for enumKey = range nameClashSet {
+				break
+			}
+
+			if s.definedEnums[clashName] {
+				// Unexpected conflict with existing name. This should not happen
+				// because clashNames are supposed to be unique.
+				return nil, fmt.Errorf("enumgen.go bug: default name %q has already been assigned during name generation, this should not happen in the algorithm; enumKey: %q; one of the entries used for name generation: %+v", clashName, enumKey, nameClashSet[enumKey])
+			}
+			uniqueNamesMap[enumKey] = clashName
+			s.definedEnums[clashName] = true
+			continue
+		}
+
+		// First, try the module name.
+		candidateUniqueNames := map[string]string{}
+		for enumKey, entry := range nameClashSet {
+			definingModNameCamelCase := yang.CamelCase(genutil.ParentModulePrettyName(entry.Node))
+			candidateUniqueNames[definingModNameCamelCase+delimiter+clashName] = enumKey
+		}
+		if addCandidateUniqueNames(candidateUniqueNames, len(nameClashSet)) {
+			continue
+		}
+
+		// FIXME(wenbli):
+		// - Remove the leading module for enumeration leaves.
+
+		// Next, try the ancestor names one by one until one succeeds
+		// or at least two of them no longer have parent entries.
+		for i := 0; ; i += 1 {
+			candidateUniqueNames = map[string]string{}
+			newNameClashSet := map[string]*yang.Entry{}
+			for enumKey, entry := range nameClashSet {
+				var candidateName string
+				if entry.Parent == nil || entry.Parent.Parent == nil {
+					// When we have reached the module-level (i.e. no
+					// grandparent), then we're out of ancestor containers to
+					// try, so just try to use the original clashName and hope
+					// everyone else can disambiguate using their parents.
+					candidateName = clashName
+					// If the name has already been used by another, then it
+					// means that there is more than one entry that hit the
+					// module-level, so we have reached a dead-end.
+					if _, ok := candidateUniqueNames[candidateName]; ok {
+						return nil, fmt.Errorf("enumgen.go: cannot resolve enumeration name clash between the following entries: %v", nameClashSet)
+					}
+					newNameClashSet[enumKey] = entry
+				} else {
+					// Try prepending the ancestor name to disambiguate.
+					candidateName = yang.CamelCase(entry.Parent.Name) + delimiter + clashName
+					newNameClashSet[enumKey] = entry.Parent
+				}
+				candidateUniqueNames[candidateName] = enumKey
+			}
+			// Update the clash set entry pointers to their next level of ancestors.
+			nameClashSet = newNameClashSet
+
+			if i == 0 {
```

Comment:
This should only apply when `compressPaths` is enabled right? i.e., we should not skip the parent directly if we're not doing path compression.

Reply:
Thanks for catching this! The uncompressed handling was missing. I've added it now, and it *should not* do any clash resolution, since uncompressed enumeration leaves are named according to their entire path.

Reply:
Great, thanks -- this handling LGTM.